sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedRleValuesReader.java

```diff
@@ -199,48 +200,51 @@ public void readBatch(
           break;
       }
       offset += n;
-      left -= n;
       currentCount -= n;
     }
+
+    state.advanceOffset(offset);
   }
 
   /**
    * Decoding for dictionary ids. The IDs are populated into `values` and the nullability is
    * populated into `nulls`.
    */
   public void readIntegers(
-      int total,
+      ParquetReadState state,
       WritableColumnVector values,
       WritableColumnVector nulls,
-      int rowId,
-      int level,
       VectorizedValuesReader data) throws IOException {
-    int left = total;
-    while (left > 0) {
+    int offset = state.offset;
+
+    while (state.hasMoreInPage(offset)) {
```

Comment:
This looks inefficient. Can we follow the old code and calculate the `left` first, then do `while (left > 0)`?

Reply:
e.g.
```
int offset = state.offset;
int left = Math.min(state.valuesToReadInPage, state. valuesToReadInBatch)
while (left > 0) {
  ...
  offset += n;
  left -= n;
}
```

Reply:
I don't have strong opinion on this. It's just that in the column index PR we'll introduce `rowId` and instead of `left` we'll  need something like `valuesLeft` and `rowsLeft`. In addition there'll be several branches where we need to update the `rowsLeft`, so it'll be something like:
```java
int offset = state.offset;
long rowId = state.rowId;
int valuesLeft = state.valuesToReadInBatch;
int rowsLeft = state.valuesToReadInPage;
while (valuesLeft > 0 && rowsLeft > 0) {
   if (...) {
    rowId += n;
    rowsLeft -= n;
  } else {
    ...
    if (toSkip > 0) {
      ...
      rowId += n;
      rowsLeft -= n;
    }
    ...    
    offset += n;
    valuesLeft -= n;
    rowId += n;
    rowsLeft -= n;
  }
}
```
so slightly more complicated.